### DISCUSSION TITLE
LFX mentorship pre-test: Basic Rust programming skills

### DISCUSSION THREAD

**Discussion by juntao** [WasmEdge/WasmEdge#2692] [General]
A lot of our LFX mentorship tasks require basic Rust programming skills. This pre-test requires you to publish serverless functions written in Rust to demonstrate your Rust skills.

**Step 1:** Make sure that you have [Rust toolchain](https://www.rust-lang.org/tools/install) and [WasmEdge](https://wasmedge.org/docs/start/install#install) installed.

 **Step 2:** [Follow the instructions](https://docs.flows.network/docs/getting-started-developer/hello-world) to create and publish a simple flow function that responds to HTTP requests. For example, you could create a `login` function that takes the username and password, and then returns the login success or failure status.

**Step 3:** [Follow the instructions](https://docs.flows.network/docs/getting-started-developer/hello-world-discord) to create and publish a Discord or Telegram bot that can answer user questions.

## Deliverables 

1. URL of the flow function-backed HTTP service. Please explain what it does and provide input / output examples.
2. Instructions to DM or chat with the Discord or Telegram bot.

**Comment by h4shk4t** [comment]
Hello @juntao 

I have completed the above steps and here are the links to the repositories:

https://github.com/h4shk4t/login-flows-network
API: https://code.flows.network/lambda/zfqeHdouBX

https://github.com/h4shk4t/flow-bot

Let me know if there are more changes to be made.

Regards
Ashutosh Srivastava

**Comment by aaryac33** [comment]
@juntao PFB my submissions and let me know the changes/corrections:
https://github.com/aaryac33/LFX-Mentorship_login
https://github.com/aaryac33/LFX-Mentorship_hello-world-discord

Thanks and regards,
Aarya

**Comment by 1awesomeJ** [comment]
@juntao,
From the flow.network docs,
there's this ```curl -X POST  https://code.flows.network/lambda/j4DPFGufPr -d "I am a Rustacean"```
could you confirm if it should be a ```POST```
I could only get a ```GET``` to work: ```curl "https://code.flows.network/lambda/3xcOkrII9F?msg=I+am+a+Rustacean" ```

**Reply by juntao** [reply]
Thank you @1awesomeJ 

You are right. The docs is wrong. Only GET works for this example. 

It would be a good exercise to figure out how to support both GET and POST. 😀

**Reply by h4shk4t** [reply]
I have made a PR for correcting the docs for enabling the POST request functionality in the program. Kindly review it.

https://github.com/flows-network/docs/pull/16

Best Regards

**Reply by 1awesomeJ** [reply]
Thanks @h4shk4t.
You're awesome.

**Comment by 1awesomeJ** [comment]
> Thank you @1awesomeJ
> 
> You are right. The docs is wrong. Only GET works for this example.
> 
> It would be a good exercise to figure out how to support both GET and POST. 😀

Certainly sir

**Comment by ana42742** [comment]
Hello @juntao! I have completed this pretest for my application:
POST request API: https://github.com/ana42742/hello-world
Discord bot: https://github.com/ana42742/hello-world-discord
Please let me know if there are any suggestions.
Cheers,
Ananya Srivastava

**Comment by 1awesomeJ** [comment]
Hi everyone,
I'm working on my discord bot, but I have issues making a ```GET``` api call. 
I used ```reqwest``` but I keep getting the issue of ```future is not 'Send'``` on its ```Response``` object. I read it could be because I'm compiling to Web Assembly. 

Then I tried ```surf``` I get the error ```"polling does not support this target OS"```

Same issues on my machine and when I try building on flow.network.

I've put in lots of hours trying to fix, so I figured "hey, why not ask for help?"

Any one has a work around? Thank you!!!

**Reply by juntao** [reply]
Hi @1awesomeJ 

You can use the http_req_wasi crate to make HTTP and HTTPS requests from Wasm. Here is an example:

https://github.com/flows-network/weather-lookup/blob/main/src/lib.rs

The reason is that HTTP networking is not yet a Wasm standard. So each runtime now has a different way to do it. We do have a reqwest_wasi crate that can support the reqwest API.

https://github.com/WasmEdge/wasmedge_reqwest_demo

Let me know if it works for you.

**Reply by 1awesomeJ** [reply]
Thank you so much sir.
I'll try it when I get back to my machine.

**Comment by 1awesomeJ** [comment]
@juntao 
The http_req_wasi crate worked! Thank you sir!!
Here are my deliverables:
Discord bot repository with an README file containing instructions on how to interact with the bot:
https://github.com/1awesomeJ/gg-discord-bot/tree/main


flow-function repsository with a README file containing instructions on how to call the HTTPS endpoint:
https://github.com/1awesomeJ/wasm-flownet-api/tree/main

URL of the flow function: https://code.flows.network/lambda/3xcOkrII9F?username=johndoe&password=foobar


Kindly let me know if there's anything outstanding on either of those sir.

Thank you.

**Reply by 1awesomeJ** [reply]
The flow function's url takes "username" and "password" as query parameters. 
I've added placeholders "johndoe" and "foobar"

**Reply by juntao** [reply]
You could put username and password into a POST body so that they are encrypted by the HTTPS.

**Reply by 1awesomeJ** [reply]
Okay sir.

Thank you so much!

**Comment by 1awesomeJ** [comment]
@juntao 
I have updated my flow function to accept a POST HTTPS request from any client through this endpoint:
https://code.flows.network/lambda/3xcOkrII9F

I have also updated the README on the repository for more detailed information on usage.

**Comment by 1awesomeJ** [comment]
So Here are my deliverables on this pre test:

Discord bot repository with an README file containing instructions on how to interact with the bot:
https://github.com/1awesomeJ/gg-discord-bot/tree/main


flow-function repsository with a README file containing instructions on how to call the HTTPS endpoint:
https://github.com/1awesomeJ/wasm-flownet-api/tree/main

URL of the flow function: https://code.flows.network/lambda/3xcOkrII9F

**Comment by YoungY620** [comment]
@juntao 
This is my solution of this pretest. Is there anything wrong with all these deliverables?

My HTTP services:
- code repo with an instruction of how to interact with it: https://github.com/casual-lab/hello-world
- URL of the API: https://code.flows.network/lambda/fhARrkEOXD

My Dscord bot designed for playing Blackjack game: 
- code repo with an documentation of how to play the game: https://github.com/YoungY620/blackjack-discord-bot
- invitation link: https://discord.com/api/oauth2/authorize?client_id=1138746375639683122&permissions=0&scope=bot

**Reply by juntao** [reply]
Hi @YoungY620 

This looks great. Thank you!

**Comment by Charles-Schleich** [comment]
Hi @juntao 
Attached are my repo's for the two pre-test deliverables. 

1. URL of the flow function-backed HTTP service. Please explain what it does and provide input / output examples.
https://github.com/Charles-Schleich/fractal-wasm-flows-network
Its a little fractal generator with some parameters to change the shape of the fractals.
Makes use of the image crate in Rust, the default request in the readme takes about 3 seconds to respond.

2. Instructions to DM or chat with the Discord or Telegram bot.
https://github.com/Charles-Schleich/flows-network-discord-bot
A little discord bot with some pre-defined responses.

Best regards
Charles

**Reply by juntao** [reply]
Thanks!

**Comment by macroexpansion** [comment]
Hi @juntao 
These are my pre-tests for the mentorship.
1. Simple Weather Flow Function
This make use of `http_req_wasi` to request weather from `OpenWeatherMap`
https://github.com/blueyellowpink/flows-simple-weather-function
2. Telegram Echo Bot
A Telegram bot that repeats messages
https://github.com/blueyellowpink/flows-chatbot

Best regards,
Andrew

**Reply by juntao** [reply]
Looks good. Thanks.

**Comment by alabulei1** [comment]
Hi @h4shk4t, @YoungY620, @ana42742, @blueyellowpink, @1awesomeJ, @aaryac33, and @Charles-Schleich,

Thank you for completing the pretest. If you have not yet submitted your application for this project, please do so before 5 PM PDT on August 25th.

**Comment by Hrushi20** [comment]
Hi @juntao,
These are my pretests for the mentorship-

1. Rust Lottery Game
https://github.com/Hrushi20/rust-lottery

2. Hello World Discord Bot
https://github.com/Hrushi20/hello-world-discord

Below is the url of the bot. 
https://discord.com/api/oauth2/authorize?client_id=1139638911182835723&permissions=0&scope=bot

**Comment by ShounakDeshmukh** [comment]
Hi @juntao 
These are my submissions for the pretests

1. Base 64 encoder decoder API 
[ShounakDeshmukh/LFX-flow-Wasmedge](https://github.com/ShounakDeshmukh/LFX-flow-Wasmedge)

Webhook Link : https://code.flows.network/webhook/3QouHzciRdtFvnsUTeT6

2. A discord bot for base 64 Encoding and Decoding
[ShounakDeshmukh/b64-discord-bot](https://github.com/ShounakDeshmukh/b64-discord-bot)

Bot URL : https://discord.com/api/oauth2/authorize?client_id=1145238972696625153&permissions=0&scope=bot

I have already applied on the LFX portal but was not aware of these pretasks as they werent mentioned on the portal 
Apologies for any delays

Regards,
Shounak

**Comment by xevor11** [comment]
Apologies for the delay, I submitted the application but did not see the pre-tests we had to do, if possible can my submission be accepted?

https://github.com/Vtewari2311/login-network-flows-pretest
